src/modules/users/users.service.ts

```
@@ -0,0 +1,94 @@
+    const page = params?.pageNumber || null
+    const size = params?.pageSize || PAGE_SIZE
+
+    if (page) {
+      const queryBuilder = await this.userRepository
+        .createQueryBuilder('user')
+        .skip((page - 1) * size)
+        .take(page * size)
+
+        if(params.firstName) {
+          queryBuilder.where('user.firstName ilike :firstName', { firstName: `%${params.firstName}%` })
+        }
+
+        if(params.lastName) {
+          queryBuilder.andWhere('user.lastName ilike :lastName', { lastName: `%${params.lastName}%`})
+        }
+
+        if(params.email) {
+          queryBuilder.andWhere('user.email ilike :email', { email: `%${params.email}%`})
+        }
+
+        if(params.nickname) {
+          queryBuilder.andWhere('user.nickname ilike :nickname', { nickname: `%${params.nickname}%`})
+        }
+
+        const [users, total] = await queryBuilder.getManyAndCount()
+
+      return {
+        data: users,
+        total
+      }
+    } else {
+      const users = await this.userRepository.find()
+
+      return {
+        data: users,
+        total: users.length
+      }
+    }
+  }
+
+  async updateUser(id: string, data: Object) {
+    const user = await this.getUser(id)
+    const newUser = user
+
+    if (user) {
+      const newUserData = Object.keys(data)
+
+      newUserData.forEach(item => {
+        if (data[item] !== user[item]) {
+          newUser[item] = data[item]
+        }
+      })
+
+      this.userRepository.save(newUser)
+    }
+  }
+
+  async updateUserProperties(id: string, props: Object) {
+    const userToBeUpdated = await this.getUser(id)
+
+    if(userToBeUpdated) {
+      const newUserData = Object.keys(props)
+      const dataToUpdate = {}
+
+      newUserData.forEach(item => {
+        if (props[item] !== userToBeUpdated[item]) {
+          dataToUpdate[item] = props[item]
+        }
+      })
+
+      this.userRepository.update(id, dataToUpdate)
+    }
```

This is the same as the update, the intent of patching functions it's to update a single property at a time, for this functions the API should accept 2 parameters, `key` and `value`, being `key` the name of the property and `value` the actual value to be updated

---

Add validation to the `key` values. E.g `id` can't be updated, therefore if the client sends `id` as the key parameter, it should throw an error, same if the parameter, requested to be updated doesn't exist